### PR TITLE
InfluxDB interface class implementation

### PR DESCRIPTION
**Pull request by Burke-Daniel**
### Info
This pull request contains a python module with the influxdb database interface class, as well as test cases. The goal of this class is to not completely wrap the functionality of the influxDB api, but rather to implement features that will be useful to us when utilizing this class elsewhere. The code is well annotated, read the docstrings to get a feel for the class, methods, and functions, and test cases demonstrate the class by utilizing the database generated by the `log-windtunnel-influx.py` script 
### Testing
See the README for how to run the test cases.

### CODE REVIEW

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: After checking this out, I created a file and upon typing import database and hovering, I get this doc string. That is super sick, I think we should do this for all modules!

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: `analysis/` , looks like you were going to continue but didn't / got distracted, I do that allllll the time

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: Wondering if the host and port should be configurable - arguments of the constructor

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py:59]: In this method, you set self.current_database = database after switching it. I think this is a little better than in the constructor above, where you set it before actually switching. In theory, without suppression of errors it will never make a difference, but in terms of logically reading the code the way it is done in the method makes more sense.

That was wordy, you feel me though?

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: See the below comment about placement of the line `self.current-database = database`. 

Also, never seen the function any(), that's cool

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: These 2 lines kind of highlight a mismatch. When logging, tags indicate metadata (host, region, etc), whereas the second call uses fields Celsius and Fahrenheit as tags.  I guess due to the implementation they are similar, but when using this test as an example/ a piece of documentation it may be misleading.

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: This function is a little bit different then I had intended. Ideally, the user can choose a tag value (i.e. profile="nicks-profile"), a measurement, and a set of fields. The function then returns all values of the selected fields that have the profile value = "nicks-profile". It is like filtering for a specific profile. If we use this design we could then have tags profileName, and testNo, so we can have numerous test runs using the same profile all in the same table and just filter for the desired ones. My idea for a tag was like a filter as opposed to a field.

The query would be something like: `"SELECT * FROM {MEASUREMENT} WHERE {tagname}=={tagvalue}"`, where tag is a dict like `{"name": "value"}`

HOWEVER, this is unnecessary if we decide to clear the tables after every test. In that case, there will only be data from one test profile at a time in the table. In this case, I still think the "tags" parameter of this function would be better named "fields".

We can call and go through this quickly to clear it up - finding it a little hard to verbalize rn

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: Im thinking this could be a little more cross platform using os.path.join(). The current path implementation seems Windows-oriented ("I think?")

`with open(os.path.join(csv_path, file_name))...` would resolve the path based on the host OS.

 Source - Denis Rouleau, Member of Technical Staff at Wind River Systems.

Comment (colton-smith) [reply Pidaq/Lib/database/database.py]: Also, maybe we could add a date to the file name - nothing long or verbose just a simple indication of the date.

Also Also, If we could prepend the file with some metadata that would also be good. Not sure if thats a good idea though because that goes against the csv format. 

Just thinking outloud- maybe a metadata file with the same name that gets created here? It could have the following fields: 
`test profile name` - easy
`date` - easy
`operator name` - this would be a neat feature, simple input box on the gui
`commands` - this would also be simple. Just accept a list of something (dict, string, whatever) that contains each command executed. This export will get called somewhere where that info is readily available anyways

What do you guys think?

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: utilized*

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py:171]: There should be 2 line breaks here because it is a top level function

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: Does this function do [1,2,3] -> ' "1", "2" , "3", ' ?

If so, there is a one line solution (not sure it needs to be a function, it is cleaner in a function though admittedly:

```
','.join([f'"{num}"' for num in lst])
```

**Review by colton-smith — CHANGES_REQUESTED** [review]
Other than a few typos and whitespace changes, there is a few things implemented a little differently than I was thinking. I was really thinking about how I would be implementing this in a telemetry thread as I reviewed it. We can jump on a call to clarify some of my points sometime soon.

Documentation and style is {fire emoji} though

File: Pidaq/Lib/database/database.py
Comment (Burke-Daniel) [reply Pidaq/Lib/database/database.py]: Ah yes, good idea with the `os.path.join`, I was thinking about that while writing and forgot to figure out how to do it. Will make that change.

I agree that this is a good idea, and creating a separate file for it makes sense cause of what you mentioned about csv formatting, it will be tidier to have a separate file.

File: Pidaq/Lib/database/database.py
Comment (Burke-Daniel) [reply Pidaq/Lib/database/database.py]: You can see that I wrote these while not 100% focused lol

File: Pidaq/Lib/database/database.py
Comment (Burke-Daniel) [reply Pidaq/Lib/database/database.py]: that'll work, gotta love a good list comprehension

File: Pidaq/Lib/database/database.py
Comment (Burke-Daniel) [reply Pidaq/Lib/database/database.py]: I think for this I may just add another parameter for fields, so then it will be:
```python
DATABASE.read_data(tags=None, fields=['Celsius', 'Fahrenheit'], measurements=['temperature'])
```
this should provide a better distinction

File: Pidaq/Lib/database/database.py
Comment (Burke-Daniel) [reply Pidaq/Lib/database/database.py]: I'm loving docstrings hahaha, we should definitely try to whereever possible

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [reply Pidaq/Lib/database/database.py]: Thats a good solution

**Comment by colton-smith** [comment]
Just realized you dont have a `__init__.py` in the package

File: Pidaq/Lib/database/database.py
Comment (Burke-Daniel) [reply Pidaq/Lib/database/database.py]: Absolutely

File: Pidaq/Lib/database/database.py
Comment (Burke-Daniel) [reply Pidaq/Lib/database/database.py:59]: Good call, I agree

File: Pidaq/Lib/database/database.py
Comment (Burke-Daniel) [reply Pidaq/Lib/database/database.py]: I am gonna have a separate function for creating a metadata file, I don't want there to be too much going on in this method

Comment (colton-smith) [reply Pidaq/Lib/database/database.py]: Good idea, may be good for that function to take a list of commands that the motor was sent and the test name

**Comment by Burke-Daniel** [comment]
@colton-smith I think I address all of your comments, ready for another look. @khpandya @afhassan @kmyali @youssef-m-ali feel free to take a look as well

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: Not sure if all `from` imports should be after all the regular `imports` or not

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: No newline after function docstrings

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: What goes in the data / tags dictionaries?

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: Raise an exception

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: os.makedirs() to create the file path if it does not exist

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: Mention that retention policy duration is minimum 1 hr and include examples for duration

**Review by colton-smith — CHANGES_REQUESTED** [review]
Few changes requested as discussed

**Comment by Burke-Daniel** [comment]
I just removed the retention policy method as it was only just wrapping a method that can be called through `self.client`, and it was messing with the tests when I called it as apparently when the default retention policy is changed, you can only query for data that was logged while that retention policy was active, or if the retention policy is specifically referenced in the query.

Other than that I think everything else is good to go. @colton-smith you can take another look if you want, I am going to wait for someone else to review as well before I go ahead and merge

File: Pidaq/Lib/database/database.py
Comment (colton-smith) [inline Pidaq/Lib/database/database.py]: Shouldn't it be one of measurement or query?

**Review by colton-smith — APPROVED** [review]
GTM

File: Pidaq/Lib/database/database.py
Comment (Burke-Daniel) [reply Pidaq/Lib/database/database.py]: Haha yes

**Pull request merged**
